agent/api/task/task.go

```diff
@@ -1213,3 +1219,20 @@ func (task *Task) AddResource(resourceType string, resource taskresource.TaskRes
 	defer task.lock.Unlock()
 	task.ResourcesMapUnsafe[resourceType] = append(task.ResourcesMapUnsafe[resourceType], resource)
 }
+
+// SetTerminalReason sets the terminalReason string and this can only be set
+// once per the task's lifecycle. This field does not accept updates.
+func (task *Task) SetTerminalReason(reason string) {
+	task.terminalReasonOnce.Do(func() {
+		seelog.Infof("Task [%s]: setting terminal reason for task [%s]", task.Arn, reason)
```

Comment:
You also need one outside this block, which records that something is `trying` to set the reason for cases in which it might be ignored.